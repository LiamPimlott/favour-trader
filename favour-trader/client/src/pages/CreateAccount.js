import React, { Component } from 'react';
import './CreateAccount.css';
import { Redirect } from 'react-router-dom';
import { Card, Form, Icon, Input, Button, message } from 'antd';
const FormItem = Form.Item;


class CreateAccount extends Component {
    constructor(){
        super();
        this.state = {
            redirect: false,
            failedAttempt: false,
            iconLoading: false,
        };
        this.renderErrorText = this.renderErrorText.bind(this);
        this.submit = this.submit.bind(this);
    }

    componentWillMount() {
        const { authService } = this.props;
        if (authService.loggedIn()) {
            this.setState({
                redirect: true,
            });
        }
    }


    submit() {
        const { authService } = this.props;
        this.props.form.validateFields((err, values) => {
            if (!err) {
                authService.signup(
                    values.firstName,
                    values.lastName,
                    values.emailAddress,
                    values.password,
                )
                    .then(res => {
                        if (res.success && res.token) {
                            this.setState({
                                redirect: true,
                            });
                        } else {
                            this.setState({
                                failedAttempt: true,
                                iconLoading: true
                            });
                            if(this.state.failedAttempt){
                                this.renderErrorText(res.message);
                            }
                        }
                    })
                    .catch(err => {
                        message.error(err.message);
                    });
            }
        })
    }

    renderErrorText(err) {
        return (
            message.error(err, 2 , ()=> {
                this.setState({
                    iconLoading: false
                });
            })
        );
    }

    render() {
        const { getFieldDecorator } = this.props.form;
        const title = <div><h5>Hone your Skills,</h5> <p>Get Started, its free!</p></div>;
        return (
            <div className={'container'}>
                {
                    (this.state.redirect) ? (<Redirect to={'/Profile'}/>) : ('')
                }
                <Card bordered title={title} className={'signup-form-card'}>
                    <Form className={'signup-form'}>
                        <FormItem>
                            {
                                getFieldDecorator('firstName', {
                                    rules: [{ required: true, message: 'Please input your first name!' },
                                            { min: 2, message: 'First name is too short!'}],
                                    validateTrigger: 'onBlur'
                                })(
                                <Input type="text"
                                   name="firstName"
                                   id="firstName"
                                   placeholder=" First Name"
                                   prefix={<Icon type="user" style={{ color: 'rgba(0,0,0,.25)' }} />}
                                />
                                )
                            }
                        </FormItem>
                        <FormItem>
                            {
                                getFieldDecorator('lastName', {
                                    rules: [{ required: true, message: 'Please input your last name!' },
                                            { min: 2, message: 'Last name is too short!'}],
                                    validateTrigger: 'onBlur'
                                })(
                                <Input type="text"
                                       name="lastName"
                                       id="lastName"
                                       placeholder=" Last Name"
                                       prefix={<Icon type="user" style={{ color: 'rgba(0,0,0,.25)' }} />}
                                />
                                )
                            }
                        </FormItem>
                        <FormItem>
                            {
                                getFieldDecorator('emailAddress', {
                                    rules: [{ required: true, message: 'Please input your E-mail address!'},
                                        {type: 'email', message: 'The input is not valid E-mail!'}],
                                    validateTrigger: 'onBlur'
                                })(
                                <Input type="email"
                                       name="email"
                                       id="email"
                                       placeholder=" Email Address"
                                       prefix={<Icon type="mail" style={{ color: 'rgba(0,0,0,.25)' }} />}
                                />
                                )
                            }
                        </FormItem>
                        <FormItem>
                            {
                                getFieldDecorator('password', {
                                    rules: [{ required: true, message: 'Please input your password!' }],
                                    validateTrigger: 'onBlur',
                                })(
                                <Input type="password"
                                       name="password"
                                       id="password"
                                       placeholder=" Password"
                                       prefix={<Icon type="lock" style={{ color: 'rgba(0,0,0,.25)' }} />}/>
                                )
                            }
                        </FormItem>
                        <Button type={'primary'} onClick={this.submit} loading={this.state.iconLoading} className={'signup-form-button'} >Submit</Button>
                        <p className={'user-agreement'}>By signing up, you agree to our <br/> <a href="">User Agreement</a></p>
                    </Form>
                </Card>
            </div>
        );
    }
}
const WrappedRegistrationForm = Form.create()(CreateAccount);
export default WrappedRegistrationForm;