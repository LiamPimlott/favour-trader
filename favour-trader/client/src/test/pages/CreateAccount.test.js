import React from 'react';
import renderer from 'react-test-renderer';
import moxios from 'moxios';
import { mount, configure } from 'enzyme';
import CreateAccount from '../../pages/CreateAccount.js';
import AuthServiceMock from '../components/AuthServiceMock.js';
import Adapter from 'enzyme-adapter-react-16';

var authServiceMock, wrapper;
//formElements
var firstNameInput, lastNameInput, emailInput, passwordInput, submitBtn;

beforeEach(function () {
    configure({ adapter: new Adapter() });
    moxios.install()
    authServiceMock = new AuthServiceMock();
    wrapper = mount(<CreateAccount authService={authServiceMock} />);
    //formElements
    firstNameInput = wrapper.find('input#firstName');
    lastNameInput = wrapper.find('input#lastName');
    emailInput = wrapper.find('input#emailAddress');
    passwordInput = wrapper.find('input#password');
    submitBtn = wrapper.find('Button[type="primary"]');
})

afterEach(function () {
    moxios.uninstall()
    authServiceMock = null;
    wrapper = null;
    firstNameInput = null;
    lastNameInput = null;
    emailInput = null;
    passwordInput = null;
})

it('renders correctly', () => {
    const tree = renderer
        .create(<CreateAccount authService={authServiceMock} />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it('verify CreateAccount Page', function () {
    expect(wrapper.contains([<h5>Hone your Skills,</h5>])).toEqual(true);
    expect(wrapper.contains([<p>Get Started, its free!</p>])).toEqual(true);
    expect(wrapper.contains([<span>Submit</span>])).toEqual(true);
    expect(firstNameInput.length).toEqual(1);
    expect(lastNameInput.length).toEqual(1);
    expect(emailInput.length).toEqual(1);
    expect(passwordInput.length).toEqual(1);
});


it('should verify error validation messages on Empty fields', function () {
    firstNameInput.simulate('blur');
    lastNameInput.simulate('blur');
    emailInput.simulate('blur');
    passwordInput.simulate('blur');
    expect(wrapper.find('.ant-form-explain').length).toBe(4);
    expect(wrapper.find('.ant-form-explain').get(0).props.children).toBe('Please input your first name!');
    expect(wrapper.find('.ant-form-explain').get(1).props.children).toBe('Please input your last name!');
    expect(wrapper.find('.ant-form-explain').get(2).props.children).toBe('Please input your E-mail address!');
    expect(wrapper.find('.ant-form-explain').get(3).props.children).toBe('Please input your password!');
});

it('should verify error validation messages on invalidate fields', function () {
    firstNameInput.simulate('change', {target: {name: 'firstName', value: 'i'}});
    firstNameInput.simulate('blur');

    lastNameInput.simulate('change', {target: {name: 'lastName', value: 'i'}});
    lastNameInput.simulate('blur');

    emailInput.simulate('change', {target: {name: 'email', value: 'i'}});
    emailInput.simulate('blur');

    expect(wrapper.find('.ant-form-explain').length).toBe(3);
    expect(wrapper.find('.ant-form-explain').get(0).props.children).toBe("First name is too short!");
    expect(wrapper.find('.ant-form-explain').get(1).props.children).toBe("Last name is too short!");
    expect(wrapper.find('.ant-form-explain').get(2).props.children).toBe("The input is not valid E-mail!");
});

it.skip('should error on invalid submission', function () {
    firstNameInput.simulate('change', { target: {name: 'firstName', value: 'Ismail'}});
    lastNameInput.simulate('change', { target: {name: 'lastName', value: 'ismail'}});
    emailInput.simulate('change', { target: {name: 'email', value: 'example@gmail.com'}});
    passwordInput.simulate('change', { target: {name: 'password', value: 'password'}});
    submitBtn.simulate('click');
}); 