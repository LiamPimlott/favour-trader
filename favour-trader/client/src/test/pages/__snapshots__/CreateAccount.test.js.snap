// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="container"
>
  
  <div
    className="ant-card signup-form-card ant-card-bordered"
  >
    <div
      className="ant-card-head"
    >
      <div
        className="ant-card-head-wrapper"
      >
        <div
          className="ant-card-head-title"
        >
          <div>
            <h5>
              Hone your Skills,
            </h5>
             
            <p>
              Get Started, its free!
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="ant-card-body"
      style={undefined}
    >
      <form
        className="ant-form ant-form-horizontal signup-form"
        onSubmit={[Function]}
      >
        <div
          className="ant-row ant-form-item"
          style={undefined}
        >
          <div
            className="ant-form-item-control-wrapper"
          >
            <div
              className="ant-form-item-control"
            >
              <span
                className="ant-form-item-children"
              >
                <span
                  className="ant-input-affix-wrapper"
                  style={undefined}
                >
                  <span
                    className="ant-input-prefix"
                  >
                    <i
                      className="anticon anticon-user"
                      style={
                        Object {
                          "color": "rgba(0,0,0,.25)",
                        }
                      }
                    />
                  </span>
                  <input
                    className="ant-input"
                    data-__field={
                      Object {
                        "name": "firstName",
                      }
                    }
                    data-__meta={
                      Object {
                        "name": "firstName",
                        "originalProps": Object {
                          "disabled": false,
                          "id": "firstName",
                          "name": "firstName",
                          "placeholder": " First Name",
                          "prefix": <Icon
                            style={
                                  Object {
                                        "color": "rgba(0,0,0,.25)",
                                      }
                            }
                            type="user"
                      />,
                          "prefixCls": "ant-input",
                          "type": "text",
                        },
                        "ref": null,
                        "rules": Array [
                          Object {
                            "message": "Please input your first name!",
                            "required": true,
                          },
                          Object {
                            "message": "First name is too short!",
                            "min": 2,
                          },
                        ],
                        "trigger": "onChange",
                        "validate": Array [
                          Object {
                            "rules": Array [
                              Object {
                                "message": "Please input your first name!",
                                "required": true,
                              },
                              Object {
                                "message": "First name is too short!",
                                "min": 2,
                              },
                            ],
                            "trigger": Array [
                              "onBlur",
                            ],
                          },
                        ],
                        "validateTrigger": "onBlur",
                        "valuePropName": "value",
                      }
                    }
                    disabled={false}
                    id="firstName"
                    name="firstName"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onKeyDown={[Function]}
                    placeholder=" First Name"
                    style={null}
                    type="text"
                    value=""
                  />
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          className="ant-row ant-form-item"
          style={undefined}
        >
          <div
            className="ant-form-item-control-wrapper"
          >
            <div
              className="ant-form-item-control"
            >
              <span
                className="ant-form-item-children"
              >
                <span
                  className="ant-input-affix-wrapper"
                  style={undefined}
                >
                  <span
                    className="ant-input-prefix"
                  >
                    <i
                      className="anticon anticon-user"
                      style={
                        Object {
                          "color": "rgba(0,0,0,.25)",
                        }
                      }
                    />
                  </span>
                  <input
                    className="ant-input"
                    data-__field={
                      Object {
                        "name": "lastName",
                      }
                    }
                    data-__meta={
                      Object {
                        "name": "lastName",
                        "originalProps": Object {
                          "disabled": false,
                          "id": "lastName",
                          "name": "lastName",
                          "placeholder": " Last Name",
                          "prefix": <Icon
                            style={
                                  Object {
                                        "color": "rgba(0,0,0,.25)",
                                      }
                            }
                            type="user"
                      />,
                          "prefixCls": "ant-input",
                          "type": "text",
                        },
                        "ref": null,
                        "rules": Array [
                          Object {
                            "message": "Please input your last name!",
                            "required": true,
                          },
                          Object {
                            "message": "Last name is too short!",
                            "min": 2,
                          },
                        ],
                        "trigger": "onChange",
                        "validate": Array [
                          Object {
                            "rules": Array [
                              Object {
                                "message": "Please input your last name!",
                                "required": true,
                              },
                              Object {
                                "message": "Last name is too short!",
                                "min": 2,
                              },
                            ],
                            "trigger": Array [
                              "onBlur",
                            ],
                          },
                        ],
                        "validateTrigger": "onBlur",
                        "valuePropName": "value",
                      }
                    }
                    disabled={false}
                    id="lastName"
                    name="lastName"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onKeyDown={[Function]}
                    placeholder=" Last Name"
                    style={null}
                    type="text"
                    value=""
                  />
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          className="ant-row ant-form-item"
          style={undefined}
        >
          <div
            className="ant-form-item-control-wrapper"
          >
            <div
              className="ant-form-item-control"
            >
              <span
                className="ant-form-item-children"
              >
                <span
                  className="ant-input-affix-wrapper"
                  style={undefined}
                >
                  <span
                    className="ant-input-prefix"
                  >
                    <i
                      className="anticon anticon-mail"
                      style={
                        Object {
                          "color": "rgba(0,0,0,.25)",
                        }
                      }
                    />
                  </span>
                  <input
                    className="ant-input"
                    data-__field={
                      Object {
                        "name": "emailAddress",
                      }
                    }
                    data-__meta={
                      Object {
                        "name": "emailAddress",
                        "originalProps": Object {
                          "disabled": false,
                          "id": "email",
                          "name": "email",
                          "placeholder": " Email Address",
                          "prefix": <Icon
                            style={
                                  Object {
                                        "color": "rgba(0,0,0,.25)",
                                      }
                            }
                            type="mail"
                      />,
                          "prefixCls": "ant-input",
                          "type": "email",
                        },
                        "ref": null,
                        "rules": Array [
                          Object {
                            "message": "Please input your E-mail address!",
                            "required": true,
                          },
                          Object {
                            "message": "The input is not valid E-mail!",
                            "type": "email",
                          },
                        ],
                        "trigger": "onChange",
                        "validate": Array [
                          Object {
                            "rules": Array [
                              Object {
                                "message": "Please input your E-mail address!",
                                "required": true,
                              },
                              Object {
                                "message": "The input is not valid E-mail!",
                                "type": "email",
                              },
                            ],
                            "trigger": Array [
                              "onBlur",
                            ],
                          },
                        ],
                        "validateTrigger": "onBlur",
                        "valuePropName": "value",
                      }
                    }
                    disabled={false}
                    id="emailAddress"
                    name="email"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onKeyDown={[Function]}
                    placeholder=" Email Address"
                    style={null}
                    type="email"
                    value=""
                  />
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          className="ant-row ant-form-item"
          style={undefined}
        >
          <div
            className="ant-form-item-control-wrapper"
          >
            <div
              className="ant-form-item-control"
            >
              <span
                className="ant-form-item-children"
              >
                <span
                  className="ant-input-affix-wrapper"
                  style={undefined}
                >
                  <span
                    className="ant-input-prefix"
                  >
                    <i
                      className="anticon anticon-lock"
                      style={
                        Object {
                          "color": "rgba(0,0,0,.25)",
                        }
                      }
                    />
                  </span>
                  <input
                    className="ant-input"
                    data-__field={
                      Object {
                        "name": "password",
                      }
                    }
                    data-__meta={
                      Object {
                        "name": "password",
                        "originalProps": Object {
                          "disabled": false,
                          "id": "password",
                          "name": "password",
                          "placeholder": " Password",
                          "prefix": <Icon
                            style={
                                  Object {
                                        "color": "rgba(0,0,0,.25)",
                                      }
                            }
                            type="lock"
                      />,
                          "prefixCls": "ant-input",
                          "type": "password",
                        },
                        "ref": null,
                        "rules": Array [
                          Object {
                            "message": "Please input your password!",
                            "required": true,
                          },
                        ],
                        "trigger": "onChange",
                        "validate": Array [
                          Object {
                            "rules": Array [
                              Object {
                                "message": "Please input your password!",
                                "required": true,
                              },
                            ],
                            "trigger": Array [
                              "onBlur",
                            ],
                          },
                        ],
                        "validateTrigger": "onBlur",
                        "valuePropName": "value",
                      }
                    }
                    disabled={false}
                    id="password"
                    name="password"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onKeyDown={[Function]}
                    placeholder=" Password"
                    style={null}
                    type="password"
                    value=""
                  />
                </span>
              </span>
            </div>
          </div>
        </div>
        <button
          className="ant-btn signup-form-button ant-btn-primary"
          onClick={[Function]}
          type="button"
        >
          <span>
            Submit
          </span>
        </button>
        <p
          className="user-agreement"
        >
          By signing up, you agree to our 
          <br />
           
          <a
            href=""
          >
            User Agreement
          </a>
        </p>
      </form>
    </div>
  </div>
</div>
`;
